Store Git commit identifier in its own object

Deiz opened the pull request:
Fixes #1640. I believe I caught all instances (within i3 itself) of the version string being compiled in as part of a static string.

Note that I've only updated i3 itself, the standalone utilities are mostly not susceptible to this by virtue of mostly being single files, though i3bar is a notable exception.

Deiz committed:
Store Git commit identifier in its own object

Fixes #1640

stapelberg commented:
How does this fix #1640 specifically? The Makefile still doesn’t have a dependency on the version itself, does it?

Deiz commented:
Since make only rebuilds objects when the source is newer than the object, there should be no difference between running `make` on commits A, B, C and running `make clean && make` on commit C.

The only problem I see in #1640 is that the version strings were separately embedded in numerous places, and so `i3 --version`, etc. would always be incorrect unless switching branches/merging commits happened to touch main.c

stapelberg commented:
My point is that the Makefile doesn’t see the git revision as an input to version.c, but it must, otherwise it won’t rebuild version.c.

Deiz committed:
Rebuild version.o when version.c or LAST_VERSION change

$(TOPDIR)/LAST_VERSION is a cached copy of common.mk's I3_VERSION var,
updated only if the two differ.

Deiz commented:
Ah, I forgot about that. I just pushed a commit that makes version.o get rebuilt when .git/HEAD or .git/index change, which is a bit too aggressive (since staging changes won't actually change the output of `git describe`) but I think that's acceptable. The only other option would be to cache the output of `git describe` in a file and compare against it, and I think that's overkill.

stapelberg commented:
> The only other option would be to cache the output of git describe in a file and compare against it, and I think that's overkill.

I actually think that’s not overkill, but the best way to avoid rebuilds when unnecessary and still rebuild whenever necessary. Can you do that please?

Deiz commented:
Updated it again, though I just realized I3_VERSION (the file) is now somewhat overloaded, because it's the canonical version for source tarballs, but not respected when within a Git repo (which of course includes cases where someone's unpacked a source tarball within a Git repo).

I think it'd be better to cache the version string as $(TOPDIR)/.last_version or similar.

Deiz commented:
Gave it a bit more thought: My proposed behaviour is to respect I3_VERSION (the file, again) if it exists, and otherwise generate/compare against LAST_VERSION, which will be the cached copy of the version string.

stapelberg commented on src/i3.mk:
```diff
@@ -92,4 +96,6 @@ install-i3: i3
-	rm -f $(i3_OBJECTS) $(i3_SOURCES_GENERATED) $(i3_HEADERS_CMDPARSER) include/loglevels.h loglevels.tmp include/all.h.pch i3-command-parser.stamp i3-config-parser.stamp i3 test.config_parser test.commands_parser src/*.gcno src/cfgparse.* src/cmdparse.*
+	rm -f $(i3_OBJECTS) $(i3_SOURCES_GENERATED) $(i3_HEADERS_CMDPARSER) include/loglevels.h loglevels.tmp include/all.h.pch i3-command-parser.stamp i3-config-parser.stamp i3 test.config_parser test.commands_parser src/*.gcno src/cfgparse.* src/cmdparse.* LAST_VERSION
+
+.INTERMEDIATE: LAST_VERSION
```
Why are you marking this file intermediate? I’ve read https://www.gnu.org/software/make/manual/html_node/Chained-Rules.html#Chained-Rules and don’t understand why that’s necessary.

Deiz replied:
I've marked it as intermediate because it's not generated when building from a source tarball, which would cause dependency failures since there isn't actually a rule for creating it.

If you'd prefer that LAST_VERSION always be created/updated (including when building from a source tarball) then it doesn't need to be intermediate.

stapelberg replied:
See my other comment. I think (without having tested it) that unconditionally creating that file would make this idea simpler.

stapelberg commented on Makefile:
```diff
@@ -22,6 +22,16 @@ include i3-dump-log/i3-dump-log.mk
 include docs/docs.mk
 include man/man.mk
 
+# Check and update $(TOPDIR)/LAST_VERSION if I3_VERSION doesn't exist
```
Can you expand on why you do that? Wouldn’t the code also work just as well if you unconditionally created LAST_VERSION?

Deiz commented:
Updated to always generate LAST_VERSION, and verified that it works properly with source tarballs as well as Git.
